File: apps/api/app/schemas/incident.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, StringConstraints
from typing import Annotated

from ..models import IncidentPriority, IncidentSeverity, IncidentStatus, EscalationStatus

Name = Annotated[str, StringConstraints(strip_whitespace=True, min_length=1, max_length=255)]
Reason = Annotated[str, StringConstraints(strip_whitespace=True, min_length=1, max_length=4000)]


class IncidentCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")
    title: Name
    description: str | None = Field(default=None, max_length=10000)
    severity: IncidentSeverity = IncidentSeverity.MEDIUM
    priority: IncidentPriority = IncidentPriority.MEDIUM
    device_id: UUID | None = None
    assigned_user_id: UUID | None = None


class IncidentUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")
    status: IncidentStatus | None = None
    priority: IncidentPriority | None = None
    severity: IncidentSeverity | None = None
    assigned_user_id: UUID | None = None
    description: str | None = Field(default=None, max_length=10000)


class IncidentAssignment(BaseModel):
    model_config = ConfigDict(extra="forbid")
    assigned_user_id: UUID


class IncidentResolution(BaseModel):
    model_config = ConfigDict(extra="forbid")
    resolution_summary: Annotated[str, StringConstraints(strip_whitespace=True, min_length=1, max_length=10000)]


class IncidentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    organization_id: UUID
    device_id: UUID | None
    assigned_user_id: UUID | None
    title: str
    description: str | None
    severity: IncidentSeverity
    priority: IncidentPriority
    status: IncidentStatus
    opened_at: datetime
    resolved_at: datetime | None
    created_by_user_id: UUID | None
    resolved_by_user_id: UUID | None
    resolution_summary: str | None
    closed_at: datetime | None
    closed_by_user_id: UUID | None
    last_escalated_at: datetime | None
    created_at: datetime
    updated_at: datetime


class IncidentPage(BaseModel):
    items: list[IncidentRead]
    total: int
    page: int
    page_size: int


class EscalationCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")
    escalation_level: int = Field(ge=1, le=100)
    reason: Reason
    assigned_to: Name | None = None


class EscalationRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    organization_id: UUID
    incident_id: UUID
    escalation_level: int
    reason: str
    status: EscalationStatus
    assigned_to: str | None
    escalated_at: datetime
    resolved_at: datetime | None
    created_at: datetime
    updated_at: datetime

File: apps/web/src/types/incident.ts
export type IncidentSeverity = 'low' | 'medium' | 'high' | 'critical'
export type IncidentPriority = 'low' | 'medium' | 'high' | 'urgent'
export type IncidentStatus = 'open' | 'in_progress' | 'resolved' | 'closed'
export type EscalationStatus = 'open' | 'acknowledged' | 'resolved'

/** apps/api/app/schemas/incident.py::IncidentRead */
export type Incident = {
  id: string
  organization_id: string
  device_id: string | null
  assigned_user_id: string | null
  title: string
  description: string | null
  severity: IncidentSeverity
  priority: IncidentPriority
  status: IncidentStatus
  opened_at: string
  resolved_at: string | null
  created_by_user_id: string | null
  resolved_by_user_id: string | null
  resolution_summary: string | null
  closed_at: string | null
  closed_by_user_id: string | null
  last_escalated_at: string | null
  created_at: string
  updated_at: string
}

/** apps/api/app/schemas/incident.py::EscalationRead */
export type Escalation = {
  id: string
  organization_id: string
  incident_id: string
  escalation_level: number
  reason: string
  status: EscalationStatus
  assigned_to: string | null
  escalated_at: string
  resolved_at: string | null
  created_at: string
  updated_at: string
}
